File: src/main/java/org/academy/internal/client/app/music/backend/MusicSource.java
package org.academy.internal.client.app.music.backend;

import net.minecraft.client.Minecraft;
import net.minecraft.resources.Identifier;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.file.Files;
import java.nio.file.Path;

public record MusicSource(Object path) {
    public static MusicSource fromIdentifier(Identifier location) {
        return new MusicSource(location);
    }

    public static MusicSource fromAbsolutePath(String absolutePath) {
        return new MusicSource(absolutePath);
    }

    public ByteBuffer getData() throws IOException {
        byte[] bytes;
        if (path instanceof Identifier location) {
            try (var stream = Minecraft.getInstance().getResourceManager().open(location)) {
                bytes = stream.readAllBytes();
            }
        } else if (path instanceof String strPath) {
            var filePath = Path.of(strPath);
            bytes = Files.readAllBytes(filePath);
        } else throw new IOException("Unsupported media source type: " + path.getClass().getName());

        var buffer = ByteBuffer.allocateDirect(bytes.length);
        buffer.put(bytes).flip();
        return buffer;
    }
}

File: src/main/kotlin/org/academy/internal/client/app/music/data/MusicSource.kt
package org.academy.internal.client.app.music.data

import net.minecraft.client.Minecraft
import net.minecraft.resources.Identifier
import java.io.IOException
import java.nio.ByteBuffer
import java.nio.file.Files
import java.nio.file.Path
import java.util.function.Supplier

data class MusicSource(val path: Any) {
    val data: ByteBuffer
        get() {
            val bytes: ByteArray
            when (path) {
                is Identifier -> Minecraft.getInstance().resourceManager.open(path).use {
                    bytes = it.readAllBytes()
                }

                is String -> bytes = Files.readAllBytes(Path.of(path))

                is Supplier<*> -> {
                    val supplied = path.get()
                    if (supplied !is ByteBuffer) throw IOException("Music source did not provide a ByteBuffer")
                    return supplied.duplicate()
                }

                else -> throw IOException("Unsupported media source type: " + path.javaClass.getName())
            }

            val buffer = ByteBuffer.allocateDirect(bytes.size)
            buffer.put(bytes).flip()
            return buffer
        }

    companion object {
        fun fromIdentifier(location: Identifier): MusicSource {
            return MusicSource(location)
        }

        fun fromAbsolutePath(absolutePath: String): MusicSource {
            return MusicSource(absolutePath)
        }

        fun fromSupplier(supplier: Supplier<ByteBuffer>): MusicSource {
            return MusicSource(supplier)
        }
    }
}
